src/core/etl/src/Flow/ETL/Pipeline.php

```diff
@@ -22,10 +24,14 @@ public function has(string $transformerClass) : bool;
 
     public function isAsync() : bool;
 
+    public function pipes() : Pipes;
+
     /**
      * @return \Generator<Rows>
      */
     public function process(FlowContext $context) : \Generator;
 
-    public function source(Extractor $extractor) : self;
+    public function setSource(Extractor $extractor) : self;
+
+    public function source() : Extractor;
```

Comment:
I had to expose the pipeline data source in order to be able to access it at the Optimizer level.